### PR TITLE
V3

### PR DESCRIPTION
**Pull request by ethanresnick**
Ongoing pr for v3

To Dos
- [ ] Finalize Adapter method return types
- [ ] Document Adapter contract, including that Adapter.find must remove duplicate resources.
- [x] Have MongooseAdapter remove duplicate included resources
- [x] Should AddToRelationshipQuery/RemoveFromRelationship query take a ResourceIdentifier[] rather than a Data<ResourceIdentifier> for their linkage? Or should document.included be a Data rather than a Resource[]?
- [x] Update ResourceIdentifier constructor to match Resource (i.e., take type and id sequentially)
- [x] Add Resource.attributes getter/setter
- [x] Figure out how/whether to have ResourceSet.data and Relationship.data getters/setters
   - Resolution: no `.data` getter/setter for v3, but rename the old, private `.data` to `._data` to leave open the possibility of adding a `.data` accessor later
- [x] Cleanup url template usage (see comment)
- [x] Deprecate and replace with a symbol `Error.isJSONAPIDisplayReady`; document the symbol
- [x] Review HTTP strategy API
- [x] Ensure it's not totally miserable to construct a Document from raw data.
- [ ] Fix deprecation notices, superFn signature in transforms

### CODE REVIEW

**Comment by carlbennettnz** [comment]
Looks like work is ongoing, but give me a shout if you want a review at some point 🙂

**Comment by ethanresnick** [comment]
> Looks like work is ongoing, but give me a shout if you want a review at some point 🙂

Thanks! A review would be wonderful. I think the code is basically production-ready now, in the sense that everything appears to work and is reasonably well tested. The big blocker for v3 at this point is that I'm trying to decide which APIs should be adjusted before pushing the semver release button and being locked into things as is. In particular, I'm thinking about things like:

- Should AddToRelationshipQuery/RemoveFromRelationship query take a `ResourceIdentifier[]` rather than a `Data<ResourceIdentifier>` for their linkage? That would be consistent with `Document.included` being a `Resource[]`, rather than a `Data<Resource>`. We could also go the other way, though, and always use a `Data` rather than an array, even when we know the contents must be conceptually plural. That has the advantage of exposing a consistent API for mapping etc., but the downside that it allows invalid values (i.e., `Data` with `isSingular: true`) through.

- Where should filtering out duplicated included resources (or included resources that are also part of primary data) happen? In Adapter.find? Before rendering the final Document? Both?

- What should the return values of the Adapter methods be? Right now, `Adapter.find` returns a tuple, which doesn't seem very scalable, while most of the other methods return a single `Data`, and the addTo/removeFromRelationship methods return idk what (`Promise<void>`?). It's a bit of an inconsistent mess, and that contract should definitely be cleaned up.

- Should the Document class ship with APIs for immutable transformations? If so, what should those be? Should they mirror the basic monadic structure of `Data` somehow? My messy mutation in https://github.com/ethanresnick/json-api/issues/105#issuecomment-354080434, as well as [this mess](https://github.com/ethanresnick/json-api/blob/v3-evolution-over-rewrite/src/controllers/API.ts#L170) makes me think some APIs for this are needed, but I haven't thought about their shape at all.

- For url templates, the ResourceTypeRegistry takes and stores them as strings, but everywhere else (i.e., the Relationship, ResourceSet, and Document types) expects to get them as functions. This should probably be consistent, with maybe every type accepting strings or functions but parsing any strings that are provided and always storing them as functions?

- The methods called `toJSON` don't actually return JSON (which would be a string), but rather return plain js objects/arrays/primitives. I'm not sure I care.

- It's probably better to replace the magic `isJSONAPIDisplayReady` property that can be added to errors with a Symbol, so there's no risk of that property showing up by accident (though that's hard to imagine). Still, a Symbol might be shorter and would make the relationship to the library clearer (since the Symbol would be imported from the lib by users attaching it to their errors).

- The APIs for the HTTP strategies and the Documentation controller haven't been reviewed at all. Might warrant some love.

So, basically, thoughts on any of the above issues, or a review of the current code with an eye towards API design decisions, would be wonderful.

I also have some design thoughts around how subtypes currently work, and around the auto-documentation. I think those should probably wait for v4 though.

**Comment by carlbennettnz** [comment]
That's a big list! Here are some initial thoughts, but I'll dig in properly later.

> Where should filtering out duplicated included resources (or included resources that are also part of primary data) happen? In Adapter.find? Before rendering the final Document? Both?

In my knex adapter, I'm grouping the included resource queries by resource type and then using `SELECT DISTINCT`, so I guess I'm doing it in the adapter. A secondary check before rendering the document would be more robust, but personally I would avoid it for performance. If you were to add it, you should also check for duplicate type/id combos in relationship arrays.

> What should the return values of the Adapter methods be? Right now, Adapter.find returns a tuple, which doesn't seem very scalable, while most of the other methods return a single Data, and the addTo/removeFromRelationship methods return idk what (Promise<void>?). It's a bit of an inconsistent mess, and that contract should definitely be cleaned up.

Fully agree that this needs some thought.

It would be possible to have `Adapter.find` only return the primary resources, and then, using the result, construct a new `Query` and call `Adapter.find` again to fetch the included resources. Includes are easily the most complicated part of my knex adapter's find method, so this simplification could potentially make writing new adapter much easier. The big drawback is, of course, that the primary and included queries would need to be run synchronously.

The new `Query` type is a big improvement for the inputs though!

> Should the Document class ship with APIs for immutable transformations? If so, what should those be? Should they mirror the basic monadic structure of Data somehow? My messy mutation in #105 (comment), as well as this mess makes me think some APIs for this are needed, but I haven't thought about their shape at all.

I can't comment on this properly without fully understanding the new datatypes, but I can say that one of the biggest pain points with the existing API has been the differences between [normal JSON:API resource structure](https://github.com/ethanresnick/json-api/blob/98756e8dc34ed70e773520a44c1f66e572b07fa5/src/types/Document.ts#L16) and what's actually exposed in transform hooks.

For example, to access the value of a to-one relationship, currently you have to check `resource.relationships.relName && resource.relationships.relName.linkage.value.id`. Its verbosity and exposure of internals (`linkage.value`) make it difficult to remember. I would really like to see resources exposed to the users of the library in structures that resemble the final JSON:API structure as closely as possible. In the case of the earlier example, this value would instead be accessed at `resource.relationships.relName.data.id`. It's still long, but I think that's what people would intuitively try.

So I guess what I'm saying is, although a fancy immutable API would be cool, allowing the user to hide a field with `delete resource.attributes.fieldName` or filter a relationships array with `resources.relationships.rel = resources.relationships.rel.filter(rel => allowed.includes(rel.id))` is much more intuitive and reduces the amount the user has to learn to use the library.

> For url templates, the ResourceTypeRegistry takes and stores them as strings, but everywhere else (i.e., the Relationship, ResourceSet, and Document types) expects to get them as functions. This should probably be consistent, with maybe every type accepting strings or functions but parsing any strings that are provided and always storing them as functions?

Yep.

I've never really liked the idea of URL templates though. Surely 99% of apps are going to use a standard pattern for all resources. I ended up writing an abstraction layer that just generates them using the resource name. Could we pick some sensible defaults and just allow overrides?

> The methods called toJSON don't actually return JSON (which would be a string), but rather return plain js objects/arrays/primitives. I'm not sure I care.

Sounds fine. Allows you to manipulate the output before stringifying if you wanted to, for some reason.

**Comment by ethanresnick** [comment]
> In my knex adapter, I'm grouping the included resource queries by resource type and then using SELECT DISTINCT, so I guess I'm doing it in the adapter.

Yeah, I think this should definitely be part of the adapter's responsibilities, because so many adapters will be able to shift this to the database, which is probably faster than us doing it in JS.

> A secondary check before rendering the document would be more robust, but personally I would avoid it for performance. 

Yeah, I agree about avoiding secondary checks (before rendering, or even right after adapter.find)  for now. I think the performance could be workable if the resources were all in a Map/hash of some sort, so that isn't bothering me too much. But I think we can always add these checks later in a non-breaking way, so no need to block v3 on that.

> I would really like to see resources exposed to the users of the library in structures that resemble the final JSON:API structure as closely as possible.

I hear this, and I think we can get a lot closer than we are now, esp. with the Resource type. A simple first step would be to add a `.attributes` getter/setter that's the same as `.attrs`. That doesn't get to your point about linkage though, which is really where the internal representation and JSON:API diverge.

> For example, to access the value of a to-one relationship, currently you have to check `resource.relationships.relName && resource.relationships.relName.linkage.value.id`

I agree that this is both long and unintuitive. We could and probably should add a getter such that reading `someRelationship.data` returns the JSON for the relationship linkage. (Right now, `relationship.data` is private and holds the backing `Data` object.) That said, part of the value of `Data` was to abstract away from JSON differences for to-one vs to-many relationships, so I don't want people to accidentally walk back into that complexity if they don't need to. E.g., 

```js
resource.relationships.relName = 
  resource.relationships.relName.filter(it => allowed.includes(it.id))
```

will work whether or not the relationship is to-one or to-many, whereas you'd need different branches if you're working with the JSON directly. So, I'm curious if just having the array helpers (filter, map, some, every) on Relationship instances -- and having those helpers work for relationships regardless of whether they're to-one or to-many (or even if they don't have data at all) -- is enough.

> It would be possible to have Adapter.find only return the primary resources, and then, using the result, construct a new Query and call Adapter.find again to fetch the included resources. 

Yeah, I think this could be nice to support in the future to make it easier to write adapters. Basically, I'm imagining some static prop on the Adapter constructor that would mean "I can't support queries with includes", in which case the library would remove the includes and issue multiple `find`s. I don't think we need this for v3 though if the only adapters are yours and mine, and you already have includes working. That's especially true because, as you point out, the sequential finds approach is much slower.

> although a fancy immutable API would be cool, allowing the user to hide a field with delete `resource.attributes.fieldName` or filter a relationships array with `resources.relationships.rel = resources.relationships.rel.filter(rel => allowed.includes(rel.id))` is much more intuitive and reduces the amount the user has to learn to use the library.

Yeah, I wasn't planning on making the Resource type immutable and, at the moment, the only types actually with an immutable API are the Query types (not sure how I feel about that). For Document in particular, the issue isn't so much around whether the API should be mutable or not; it's more just whether there's the right API to make common things convenient. Thinking about it more, I'm not exactly sure what would be useful to add, so maybe there is. One option would be to add a `Document.data` getter like the `Relationship.data` getter proposed above, to return the JSONified version of the primary data. That would replace the `resultDoc.primary && resultDoc.primary.toJSON() === null` check from my earlier comment with just `resultDoc.data === null`, which is kinda nice. Like the other getter, though, I'm not sure yet if it's useful enough to justify inclusion.

>  maybe every type accepting strings or functions but parsing any strings that are provided and always storing them as functions?

On second thought, we probably want to store the template string somewhere to use in the documentation, so I want to think about this more.

> Surely 99% of apps are going to use a standard pattern for all resources. I ended up writing an abstraction layer that just generates them using the resource name. Could we pick some sensible defaults and just allow overrides?

Agreed that most apps will use the same thing everywhere. I'm not sure why you needed an abstraction layer, though, as I'm pretty sure the resource type name gets passed to the template as a parameter (right?). If so, picking sensible defaults should be as simple as adding some templates to [these global resource type description defaults](https://github.com/ethanresnick/json-api/blob/v3-evolution-over-rewrite/src/ResourceTypeRegistry.ts#L15), which the library uses when the user provides nothing.

**Comment by ethanresnick** [comment]
> If so, picking sensible defaults should be as simple as adding some templates to these global resource type description defaults, which the library uses when the user provides nothing.

Actually, to make that work we'd at least have to know the API's base url.

**Comment by carlbennettnz** [comment]
> So, I'm curious if just having the array helpers (filter, map, some, every) on Relationship instances -- and having those helpers work for relationships regardless of whether they're to-one or to-many (or even if they don't have data at all) -- is enough.

Yeah, good point. Still adds to the API surface area, but having attributes and data getters would fix most of the issues I've had.

> as you point out, the sequential finds approach is much slower.

Definitely true for Mongo, not so sure for SQL databases though. I haven't benchmarked, but my knex adapter will send off a big, complicated query for the primary resources (supporting filters across to-many relationships needs a bunch of joins and subqueries) and then effectively run it again in parallel as a subquery in the WHERE clause of each included resources query. It's probably still faster than waiting for the primary query, and that's why I did it this way, but I'm not sure it's a clear win in all cases.

> I don't think we need this for v3 though if the only adapters are yours and mine, and you already have includes working.

👍 

> I'm not sure why you needed an abstraction layer, though, as I'm pretty sure the resource type name gets passed to the template as a parameter (right?).

TIL that exists 🙃  Makes my point mostly moot, though defaults would still be nice. Does JSON:API specify that links need to be a full URL, or would just the path be fine?

**Comment by ethanresnick** [comment]
@carlbennettnz I've fixed a couple of the issues discussed above and added a checklist to the OP.

> Does JSON:API specify that links need to be a full URL, or would just the path be fine?

It's [a bit contested](https://github.com/json-api/json-api/issues/898) iirc, but full urls are definitely safer (and are probably easy enough with the inherited templates).

**Comment by carlbennettnz** [comment]
Hey Ethan, sorry for going dark on you. I was midway through testing the polymorphic resource types API design with Ember Data when I got sick and ended up in hospital. Nothing too serious and I’m fine now, I just had a big backlog of stuff to do once I got out, meaning this fell down the priority list a little.

Good news though: I have a new greenfield project coming up that I can use to dog food v3! Development should be getting underway later this week. I’ll let you know how it goes, and I’ll also try to finish off experimenting with the Ember Data polymorphic stuff.

**Comment by ethanresnick** [comment]
Hey Carl,

Sorry to hear that you were in the hospital, and I'm super glad you're out now. 

I can't wait to hear how using v3 goes as well. I've also been using it on a new, much more complicated API, and it's held up pretty well so far. (The custom query machinery in particular has proven invaluable). 

One thing that's been tricky is that requests for truly custom endpoints often need to be mapped to multiple queries, not just one, which is something I've been thinking about doing for [a long time](https://github.com/ethanresnick/json-api/blob/v3-wip/src/types/index.ts#L376). I might try to finally add that in the coming days, although, if I do, I'll open an issue first to discuss how.

**Comment by carlbennettnz** [comment]
I'm trying to get my [Knex adapter](https://github.com/carlbennettnz/json-api-knex-adapter/blob/v3/src/knex-adapter.ts) working with v3. I'm 99% sure I'm returning the same stuff as the Mongoose adapter, but I'm getting errors in the json-api lib.

Is there anything still holding up adding adapter method return types? That should pick up the issue, if it is a problem in my code.

<details>
<summary>Stack trace, if it matters</summary>

```
Errors converted to json-api Result TypeError: this.value.data.map is not a function
    at Data.<anonymous> (/Users/cbnz/Projects/json-api-knex-adapter/node_modules/json-api/build/src/types/Generic/Data.js:31:73)
    at Generator.next (<anonymous>)
    at /Users/cbnz/Projects/json-api-knex-adapter/node_modules/json-api/build/src/types/Generic/Data.js:7:71
    at new Promise (<anonymous>)
    at __awaiter (/Users/cbnz/Projects/json-api-knex-adapter/node_modules/json-api/build/src/types/Generic/Data.js:3:12)
    at Data.flatMapAsync (/Users/cbnz/Projects/json-api-knex-adapter/node_modules/json-api/build/src/types/Generic/Data.js:30:16)
    at Document.<anonymous> (/Users/cbnz/Projects/json-api-knex-adapter/node_modules/json-api/build/src/types/Document.js:119:60)
    at Generator.next (<anonymous>)
    at /Users/cbnz/Projects/json-api-knex-adapter/node_modules/json-api/build/src/types/Document.js:7:71
    at new Promise (<anonymous>)
    at __awaiter (/Users/cbnz/Projects/json-api-knex-adapter/node_modules/json-api/build/src/types/Document.js:3:12)
    at Document.transform (/Users/cbnz/Projects/json-api-knex-adapter/node_modules/json-api/build/src/types/Document.js:83:16)
    at Object.<anonymous> (/Users/cbnz/Projects/json-api-knex-adapter/node_modules/json-api/build/src/controllers/API.js:274:20)
    at Generator.next (<anonymous>)
    at /Users/cbnz/Projects/json-api-knex-adapter/node_modules/json-api/build/src/controllers/API.js:7:71
    at new Promise (<anonymous>)
    at __awaiter (/Users/cbnz/Projects/json-api-knex-adapter/node_modules/json-api/build/src/controllers/API.js:3:12)
    at Object.transformDoc (/Users/cbnz/Projects/json-api-knex-adapter/node_modules/json-api/build/src/controllers/API.js:272:12)
    at Object.<anonymous> (/Users/cbnz/Projects/json-api-knex-adapter/node_modules/ramda/src/internal/_createPartialApplicator.js:8:17)
    at Object.transformDocument (/Users/cbnz/Projects/json-api-knex-adapter/node_modules/ramda/src/internal/_arity.js:14:19)
    at /Users/cbnz/Projects/json-api-knex-adapter/node_modules/json-api/build/src/controllers/API.js:173:33
    at Generator.next (<anonymous>)
    at /Users/cbnz/Projects/json-api-knex-adapter/node_modules/json-api/build/src/controllers/API.js:7:71
    at new Promise (<anonymous>)
    at __awaiter (/Users/cbnz/Projects/json-api-knex-adapter/node_modules/json-api/build/src/controllers/API.js:3:12)
    at transformResultDocument (/Users/cbnz/Projects/json-api-knex-adapter/node_modules/json-api/build/src/controllers/API.js:171:57)
    at /Users/cbnz/Projects/json-api-knex-adapter/node_modules/ramda/src/internal/_pipeP.js:5:16
    at <anonymous>
```

</details>

**Comment by ethanresnick** [comment]
Ok, what do you think about these for the return types:

```
// for find(), use same as today but put the results in an object for clarity in autocompletes
// and easier destructuring if caller is only interested in some values.
find: Promise<{ primary: Data<Resource>, included: Data<Resource>, collectionSize: number | null }>

// for create/update, put the result in an object too.
// adds uniformity but is more so in case we want to return other information later too.
create: Promise<{ created: Data<Resource> }>
update: Promise<{ updated: Data<Resource> }>

// similar to above, but allow { deleted: void } as it often won't be necessary/
// make sense for the adapter to return the deleted resources
delete: Promise<{ deleted: void | Data<Resource> }>

// similar to delete in the rationale for allowing void. support returning 
// the old state of the relationship as well, because it's might be marginally 
// interesting for the caller to know whether, on add, one of the added items 
// was already in the relationship or, on delete, that it was already missing.
addToRelationship: Promise<{ new: Relationship | void, old: Relationship | void }>
removeFromRelationship: Promise<{ new: Relationship | void, old: Relationship | void }>
```

**Comment by carlbennettnz** [comment]
A few points on the interface:

* Could we elevate the single/collection logic out of the adapters? It would allow us to avoid repeating the same logic in each adapter, and remove the need for `Data<Resource>` instead of just `Resource[]`. The query could just be a filter (id = x) and a limit of 1.
* For the find method, should `collectionSize` be optional?
* Personally I'd prefer `null` over `void`/`undefined`. I think of `undefined` as meaning you forgot to set a value, whereas `null` is explicitly "there's not a value here". I think that's general practice, but I might be wrong.
* With the relationship methods, I would avoid `new` as the key, just because it's a JS keyword. Maybe `before` and `after`?

<details>
<summary>Value of `primary` that produces the error. Definitely no empty `Data` 🤔 </summary>

```
Data {
  value:
   { data:
      [ Resource {
          _type: 'posts',
          _id: '000000000000000000000001',
          _attrs: { title: 'Post 1', date: 2017-06-01T00:00:00.000Z },
          _relationships:
           { author:
              Relationship {
                links: {},
                _data:
                 Data {
                   value:
                    { data:
                       [ ResourceIdentifier { type: 'posts', id: '000000000000000000000001' } ],
                      isSingular: true } },
                owner: { type: 'posts', id: '000000000000000000000001', path: 'author' } },
             tags:
              Relationship {
                links: {},
                _data:
                 Data {
                   value:
                    { data:
                       [ ResourceIdentifier { type: 'posts', id: '000000000000000000000001' },
                         ResourceIdentifier { type: 'posts', id: '000000000000000000000002' } ],
                      isSingular: false } },
                owner: { type: 'posts', id: '000000000000000000000001', path: 'tags' } },
             comments:
              Relationship {
                links: {},
                _data:
                 Data {
                   value:
                    { data:
                       [ ResourceIdentifier { type: 'posts', id: '000000000000000000000001' },
                         ResourceIdentifier { type: 'posts', id: '000000000000000000000002' },
                         ResourceIdentifier { type: 'posts', id: '000000000000000000000003' } ],
                      isSingular: false } },
                owner:
                 { type: 'posts',
                   id: '000000000000000000000001',
                   path: 'comments' } } },
          _meta: {} } ],
     isSingular: true } }
```

</details>

**Comment by carlbennettnz** [comment]
Fixed the problem. My `primary` was fine. I was doing this:

    return [ primary, { included } ]

Should have been this:

    return [ primary, included ]

For some reason I had been looking at [this commented out line](https://github.com/ethanresnick/json-api/blob/v3-evolution-over-rewrite/src/db-adapters/Mongoose/MongooseAdapter.ts#L229) in the Mongoose adapter and ignoring the actual return type.

**Comment by carlbennettnz** [comment]
I'm getting TypeScript errors when trying to build a project that uses v3. Any idea what's causing them?

Example repo: https://github.com/carlbennettnz/json-api-ts-error

Without mongoose:

```
$ npx tsc
node_modules/json-api/build/src/controllers/API.d.ts(2,23): error TS2688: Cannot find type definition file for 'ramda'.
node_modules/json-api/build/src/db-adapters/Mongoose/MongooseAdapter.d.ts(1,23): error TS2688: Cannot find type definition file for 'mongoose'.
node_modules/json-api/build/src/db-adapters/Mongoose/MongooseAdapter.d.ts(2,26): error TS2307: Cannot find module 'mongodb'.
node_modules/json-api/build/src/db-adapters/Mongoose/MongooseAdapter.d.ts(3,27): error TS2307: Cannot find module 'mongoose'.
node_modules/json-api/build/src/db-adapters/Mongoose/MongooseAdapter.d.ts(5,33): error TS2307: Cannot find module 'mongoose'.
```

With mongoose:

```
$ npx tsc
node_modules/json-api/build/src/controllers/API.d.ts(2,23): error TS2688: Cannot find type definition file for 'ramda'.
node_modules/json-api/build/src/db-adapters/Mongoose/MongooseAdapter.d.ts(1,23): error TS2688: Cannot find type definition file for 'mongoose'.
```

**Comment by ethanresnick** [comment]
> Could we elevate the single/collection logic out of the adapters? It would allow us to avoid repeating the same logic in each adapter, and remove the need for Data<Resource> instead of just Resource[]. The query could just be a filter (id = x) and a limit of 1.

I agree that the logic is gonna be the same in each adapter, which is a bit annoying. Maybe the library can provide some utility methods that all adapters can use to handle this? I think that's better than moving it out of the adapter...  conceptually, I don't like the thought that the code interpreting the adapter result needs to inspect/know about the query in order to build a proper Result.

> For the find method, should collectionSize be optional?

Yes, I like that. And then we can remove null from the type too, so it's just `{ collectionSize?: number, ..... }`.

> Personally I'd prefer null over void/undefined. I think of undefined as meaning you forgot to set a value, whereas null is explicitly "there's not a value here". I think that's general practice, but I might be wrong.

I think of `null` vs `undefined` similarly. For all the properties with `void` in their type, I meant them as optional properties, which I think is more appropriate than `null` (the adapter just isn't providing the deleting resources, or the old relationship state, even though they do exist conceptually/aren't affirmatively null). I should've written them as `{ x?: T}` rather than `{ x: void | T }`.

> With the relationship methods, I would avoid new as the key, just because it's a JS keyword. Maybe before and after?

Good call.

That leaves these as the updated types:

```ts
find: Promise<{ primary: Data<Resource>, included: Data<Resource>, collectionSize?: number }>
create: Promise<{ created: Data<Resource> }>
update: Promise<{ updated: Data<Resource> }>
delete: Promise<{ deleted?: Data<Resource> }>
addToRelationship: Promise<{ before?: Relationship, after?: Relationship }>
removeFromRelationship: Promise<{ before?: Relationship, after?: Relationship }>
```

Anything else?

**Comment by ethanresnick** [comment]
> I'm getting TypeScript errors when trying to build a project that uses v3. Any idea what's causing them?

The ramda thing is just a mistake... I was putting its typings in devDependencies, whereas they should've been in normal dependencies. That's fixed now. For the mongoose stuff, I'm not sure what to do exactly. I guess I could move `@types/mongoose` to `dependencies` instead of `devDependencies`, but there's still gonna be the `Cannot find module 'mongoose'` error for people who don't install mongoose. So, I think the best fix -- though I'm open to other ideas -- is for Typescript projects that depend on json-api but don't install mongoose to add ambient module declarations for mongoose/mongodb (i.e., in a `.d.ts` file, `declare module "mongoose"; declare module "mongodb";`). I'm pretty sure that should work.

**Comment by ethanresnick** [comment]
> In v2 this could be set with like the 5th or 6th arg to the Error constructor, but it seems that's been removed. Was that intentional?

I'm pretty sure there was never a way to set `source` — instead there was an argument for a totally-made-up `paths` property, which I removed [here](https://github.com/ethanresnick/json-api/commit/65de7e5b45ee69b2c3f3cd9d7846a4e49a34b3de). There definitely should be a way to set source, though, so I'll add that now.

**Comment by carlbennettnz** [comment]
> I think that's better than moving it out of the adapter... conceptually, I don't like the thought that the code interpreting the adapter result needs to inspect/know about the query in order to build a proper Result.

That's fine 👍 

> That leaves these as the updated types

`included` can't be singular and it can be `void`, so should the type be `included?: Resource[]`? There is an argument to leave it as-is and just use an empty `Data` when there are no included resources, but it should probably be consistent with the delete method.

Curious, what are the resources returned from `delete()` used for? They weren't returned in v2, right?

> but there's still gonna be the Cannot find module 'mongoose' error for people who don't install mongoose

I still think #142 is the way to go here.

> I'm pretty sure there was never a way to set `source`

Yep, you're right. Know I think about it, it was a feature I added in my fork.

---

One other thing: Do we need a `doQuery` method? The way it is now, `find()`, `create()`, `update()`, etc. are never called directly and so they probably shouldn't be in the interface at all. Is there a situation where you need direct control over all query types?

**Comment by carlbennettnz** [comment]
How to determine whether or not the result should be singular is a bit inconsistent. Could it be standardised to `query.singular` in all cases?

| Method | How to check |
| ------- | ----------------- |
| Find      | `query.singular` |
| Create  | `query.records.isSingular` |
| Update | `query.patch.isSingular` |
| Delete | `query.singular` |

**Comment by ethanresnick** [comment]
> should the type be `included?: Resource[]`?

Yes. I've updated my prior post. I meant to keep the type for included as-is, but forgot that it was a `Resource[]` and not a `Data<Resource>`. Using an array makes perfect sense.

> Curious, what are the resources returned from delete() used for? They weren't returned in v2, right?

They probably won't be used for anything. The thought was more that, if the adapter happens to need to fetch them anyway, it might as well return them. This could probably just be `Promise<{}>` for now though.

> I still think #142 is the way to go here.

I can see that. I still want to think about it more, but that could make sense.

> One other thing: Do we need a doQuery method? 

I'm not sure sure...

> Is there a situation where you need direct control over all query types?

Can you elaborate on what you mean by this?

> Could it be standardised to `query.singular` in all cases?

I see the initial appeal, but I don't think it really makes sense... `query.[records|patch].isSingular` is going to need to continue to exist, because it comes from `Data`, and is used for serialization throughout. So, adding a `query.singular` in create/update queries would just introduce a second source of "isSingular" information that would have to be set, and that could potentially disagree with the former.

**Comment by ethanresnick** [comment]
@carlbennettnz FYI, just published an update with support for `APIError.source`

**Comment by carlbennettnz** [comment]
> > Curious, what are the resources returned from delete() used for? They weren't returned in v2, right?
>
> They probably won't be used for anything. The thought was more that, if the adapter happens to need to fetch them anyway, it might as well return them. This could probably just be Promise<{}> for now though.

My adapter isn't currently fetching them, although that may change when I implement subtypes. `Promise<{}>` or `Promise<void>` would be good.

> > Is there a situation where you need direct control over all query types?
>
> Can you elaborate on what you mean by this?

So I guess the advantage of having the adapters implement `doQuery()` is that, if needed, they would be able to easily apply common logic or checks or whatever for all queries, regardless of type. The most realistic example I can think of would be error handling. `doQuery()` would be a nice place to wrap all calls to the query handlers in a try/catch block, and throwing a slightly more useful error.

That said, you could still easily implement the same sort of logic without `doQuery()` and I would probably rather see a slightly smaller, simpler adapter interface.

> I see the initial appeal, but I don't think it really makes sense... query.[records|patch].isSingular is going to need to continue to exist, because it comes from Data, and is used for serialization throughout. So, adding a query.singular in create/update queries would just introduce a second source of "isSingular" information that would have to be set, and that could potentially disagree with the former.

Yeah, good point. Perhaps we could just standardise the key name, by making it `query.isSingular` then?

> FYI, just published an update with support for APIError.source

Awesome, thanks for that :)

**Comment by carlbennettnz** [comment]
> Servers MUST send all JSON API data in response documents with the header Content-Type: application/vnd.api+json without any media type parameters.

Seems we're currently sending `application/json; charset=utf-8` if the `Accept` header isn't set.

**Comment by ethanresnick** [comment]
> Seems we're currently sending application/json; charset=utf-8 if the Accept header isn't set.

This is technically valid (and it's just not "JSON API data" at that point). See discussion in https://github.com/json-api/json-api/pull/703

If this is a problem for some reason, though, I think I'd be open to changing the Content-Type in this case to `application/vnd.api+json` (which HTTP also allows), but I suspect the latter is more likely to cause casual clients problems (e.g., in-browser json formatters, etc., which can be useful during development/debugging) than leaving things as is.

**Comment by ethanresnick** [comment]
Also, I just realized I never responded to your earlier comment....

> `Promise<{}>` or `Promise<void>` would be good.

Promise<{}> it is. I've updated my earlier comment.

> So I guess the advantage of having the adapters implement doQuery() is that, if needed, they would be able to easily apply common logic or checks or whatever for all queries, regardless of type.

That's an advantage, but the other advantage is that it gives users of the adapter a way to throw any query they have at it, without knowing/switching on the query's type and calling the proper method. The API controller uses this, as the various make-query functions can return queries of different types, but also a couple custom endpoints I have in one of my APIs call it. So, I think it's nicer to keep the `switch` centralized in Adapter.makeQuery, rather than scattering it around in consumers. As with the single/collection logic we mentioned earlier, though, it would eventually be good to put this switch in a helper function or a base Adapter class/mixin or similar so not all adapters have to reimplement it.

> Perhaps we could just standardise the key name, by making it query.isSingular then?

Will do.

**Comment by carlbennettnz** [comment]
I'm still having a bit of trouble getting an app using `json-api@3.0.0-beta.24` to compile with TypeScript. Even with `mongoose` and `@types/mongoose` installed, I still get this

    node_modules/json-api/build/src/db-adapters/Mongoose/MongooseAdapter.d.ts(1,23): error TS2688: Cannot find type definition file for 'mongoose'.

Adding an app.d.ts file with this in it only makes more errors, and doesn't stop the one above.

```ts
declare module 'mongoose'
declare module 'mongodb'
```

Any ideas?

**Comment by carlbennettnz** [comment]
2.8.1

**Comment by ethanresnick** [comment]
Cool. And what's the error exactly? One thing I can think of is that you may need to install `@types/mongodb` too. All this install trouble is convincing me that you may be right about #142.

Also, any thoughts on #160? Would those changes pose any serious problems?

**Comment by carlbennettnz** [comment]
What's the story with the `req.host` deprecation? I was gonna write a PR to fix it, but you've disabled the tslint warning. Is it more complicated that just swapping `req.host` for `req.hostname`?

https://github.com/ethanresnick/json-api/blob/572945dc1b811db61e2e8b30d82cfbcd25479c86/src/http-strategies/Express.ts#L52

**Comment by ethanresnick** [comment]
Ok, I'm going to merge this into master and open a new PR for the remaining changes. The old CircleCI status check isn't working here for some reason and it's just a bit hard to follow.

**Comment by ethanresnick** [comment]
> What's the story with the req.host deprecation? I was gonna write a PR to fix it, but you've disabled the tslint warning. Is it more complicated that just swapping req.host for req.hostname?

The issue is that we actually want the host (i.e., with the port), and not just the hostname. In express 4,  there's no built-in way to get this. I could try to write code to get the port and build the host properly, but I don't think that's worth it because: 1) for security, users should really be providing the host manually anyway (with the http-strategy option), and if lack of port support forces some users to do that, that's fine; but then 2) once express 5 comes out, this will just start behaving properly, and I don't want to have to update the code then to not double add the port.
